File: OpenGSGLibrary/WorldMap/Projection.vb
'    OpenSGSGLibrary is an open-source library for Grand Strategy Games
'    Copyright (C) 2020  Torben Kneesch
'
'    This program is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    This program is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.
Imports System.Device.Location

Namespace Map

    ''' <summary>
    ''' Base class for map projections.
    ''' Provides transformation functions between map pixel coordinates and global coordinates.
    ''' </summary>
    Public MustInherit Class Projection

        Public MustOverride Function getMapCoordinates(globePoint As GeoCoordinate) As Tuple(Of Double, Double)

        Public MustOverride Function getGlobeCoordinates(mapPoint As Tuple(Of Double, Double)) As GeoCoordinate

    End Class

End Namespace

File: OpenGSGLibrary/WorldMap/Projection.cs
using OpenGSGLibrary.SystemDeviceLocation;

namespace OpenGSGLibrary.WorldMap
{
    /// <summary>
    /// Base projection for converting between globe coordinates (latitude/longitude)
    /// and 2D map coordinates used by the rendering system.
    /// Implementations should provide forward and inverse transformations.
    /// </summary>
    public abstract class Projection
    {
        /// <summary>
        /// Converts a geographic coordinate to map (x,y) coordinates.
        /// </summary>
        /// <param name="globePoint">Geographic point (latitude/longitude).</param>
        /// <returns>Tuple with map X and Y coordinates.</returns>
        public abstract Tuple<double, double> GetMapCoordinates(GeoCoordinate globePoint);

        /// <summary>
        /// Converts map (x,y) coordinates back to geographic coordinates.
        /// </summary>
        /// <param name="mapPoint">Tuple with map X and Y coordinates.</param>
        /// <returns>Geographic coordinate (latitude/longitude).</returns>
        public abstract GeoCoordinate GetGlobeCoordinates(Tuple<double, double> mapPoint);
    }
}
